### PR TITLE
Add support for custom derivatives in jax.experimental.callback

### PR DESCRIPTION
**Pull request by sharadmv**
This PR enables `callback` to rewrite the bodies of `custom_derivatives` primitives. 

Any suggestions on how to test the custom_call_jaxpr primitives? I'm not sure how/why they are created normally.

### CODE REVIEW

**Comment by mattjj** [comment]
> Any suggestions on how to test the custom_call_jaxpr primitives? I'm not sure how/why they are created normally.

They're created when the custom derivative calls are staged out. That is, `custom_jvp_call` never appears in a jaxpr and is just a Python tracing primitive; when tracing a `custom_jvp` function to a jaxpr by e.g. `lax.scan`, we [bind a `custom_jvp_call_jaxpr`](https://github.com/google/jax/blob/8d3b4ac2f36761513ae132b59180e9770fe34922/jax/custom_derivatives.py#L221). So unlike `xla_call_p`, which is both a Python tracing primitive and a jaxpr primitive, we have pairs of primitives `custom_jvp_call_p`/`custom_jvp_call_jaxpr_p` and `custom_vjp_call_p`/`custom_vjp_call_jaxpr_p`.

TL;DR test them by using `scan`, like in [these tests](https://github.com/google/jax/blob/8d3b4ac2f36761513ae132b59180e9770fe34922/tests/api_test.py#L3041-L3128).

It might be worth adding a test like that, just for basic functionality, before we land-and-iterate!

**Comment by sharadmv** [comment]
I added a test, but left a TODO to enable rewriting thunks (because I'm not sure of the best way to do that). If it's easy, I can add it to the PR, but if not, we can iterate on it later. WDYT?

**Comment by sharadmv** [comment]
Thunks should be working now!

**Review by mattjj — APPROVED** [review]
Awesome! Thanks for doing this.
